### PR TITLE
Big refactor of BigQuery query handling

### PR DESCRIPTION
**Pull request by jskeet**
BigqueryResult has been replaced by BigqueryQueryJob.
That's basically a wrapper over GetQueryResultsResponse.
The normal workflow would be:

- Create a query job (potentially by calling ExecuteQuery)
- Poll for completion
- Either use GetRows() to get all results lazily, or
  GetResultSet() to eagerly fetch a specific number of results,
  with a page token for getting the next one.

This definitely needs a lot of documentation - and some more tests -
but I'd like to get the first pass at least reviewed first. We may
want to:

- Rename BigqueryQueryJob
- Add helper methods to make it really easy to just get a result set
  for a query, including specifying the options automatically

### CODE REVIEW

**Comment by jskeet** [comment]
@SurferJeffAtGoogle Do you have any BigQuery samples yet? If so, I'd be happy to help show you what they'd look like after this refactoring, as an aid to judging the usability.

**Comment by jskeet** [comment]
cc @jsimonweb - not sure why I can't assign it to you as well...

File: apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryClient.cs [hunk 1]
```diff
@@ -65,6 +65,7 @@ public abstract partial class BigqueryClient
         /// The credentials are scoped as necessary.
         /// </remarks>
         /// <param name="credential">Optional <see cref="GoogleCredential"/>.</param>
+        /// <param name="projectId">The ID of the project containing the BigQuery data. Must not be null.</param>
```
Comment (chrisdunelm) [inline apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryClient.cs]: Param xmldocs in the wrong order

File: apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryClient.cs [hunk 2]
```diff
@@ -81,6 +82,7 @@ public abstract partial class BigqueryClient
...
```
Comment (chrisdunelm) [inline apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryClient.cs]: Also in the wrong order

File: apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryClientImpl.Queries.cs [hunk 3]
```diff
@@ -69,19 +69,38 @@ public override BigqueryJob CreateQueryJob(string sql, CreateQueryJobOptions opt
             return new BigqueryJob(this, job);
         }
 
-        /// <inheritdoc />
-        public override BigqueryResult GetQueryResults(JobReference jobReference, GetQueryResultsOptions options = null)
+        /// <inheritdoc />
```
Comment (chrisdunelm) [inline apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryClientImpl.Queries.cs]: nit: trailing spaces

File: apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryClientImpl.Queries.cs [hunk 4]
```diff
@@ -69,19 +69,38 @@ public override BigqueryJob CreateQueryJob(string sql, CreateQueryJobOptions opt
...
+            for (long i = 0; i < maxRequests && DateTimeOffset.UtcNow < deadline; i++)
             {
-                var request = Service.Jobs.GetQueryResults(jobReference.ProjectId, jobReference.JobId);
-                options?.ModifyRequest(request);
-                return request;
-            };
-            var firstResponse = requestProvider().Execute();
-            return new BigqueryResult(this, firstResponse, requestProvider);
+                var job = GetQueryJob(jobReference, options);
+                Console.WriteLine($"{DateTime.Now:HH:mm:ss} Polled... completed? {job.Completed}");
```
Comment (chrisdunelm) [inline apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryClientImpl.Queries.cs]: Remove this writeline?

File: apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryClientImpl.Queries.cs [hunk 5]
```diff
@@ -69,19 +69,38 @@ public override BigqueryJob CreateQueryJob(string sql, CreateQueryJobOptions opt
...
 
-        /// <inheritdoc />
-        public override BigqueryResult GetQueryResults(JobReference jobReference, GetQueryResultsOptions options = null)
+        /// <inheritdoc />        
+        public override BigqueryQueryJob PollQueryUntilCompleted(JobReference jobReference, GetQueryResultsOptions options = null, PollJobOptions pollOptions = null)
         {
             GaxRestPreconditions.CheckNotNull(jobReference, nameof(jobReference));
+            pollOptions?.Validate();
+
+            DateTimeOffset? deadline = pollOptions?.GetEffectiveDeadline() ?? DateTimeOffset.MaxValue;
```
Comment (chrisdunelm) [inline apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryClientImpl.Queries.cs]: Why is this a DateTimeOffset, rather than a DateTime?
Can't we simply insist all datetime's are UTC?

File: apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryQueryJob.cs [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+using Google.Api.Gax.Rest;
+using Google.Apis.Bigquery.v2.Data;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Google.Bigquery.V2
+{
+    // TODO: Rename this to avoid the double "query".
```
Comment (chrisdunelm) [inline apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryQueryJob.cs]: Yes, I was thinking this too.

File: apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryQueryJob.cs [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+
+        internal BigqueryQueryJob(BigqueryClient client, QueryResponse response, ExecuteQueryOptions options)
+            : this(client, ConvertQueryResponse(GaxRestPreconditions.CheckNotNull(response, nameof(response))), options?.ToGetQueryResultsOptions())
+        {
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="pollOptions"></param>
```
Comment (chrisdunelm) [inline apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryQueryJob.cs]: Missing doc contents

File: apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryQueryJob.cs [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+        /// <returns></returns>
+        public BigqueryQueryJob PollUntilCompleted(PollJobOptions pollOptions = null) =>
+            Completed ? this : _client.PollQueryUntilCompleted(JobReference, _options, pollOptions: pollOptions);
+
+        // Note: this is a method rather than a property so that later we can add the async version.
+
+        /// <summary>
+        /// Returns a lazy enumeration of all the rows in the query (from any originally-specified page token onwards),
```
Comment (chrisdunelm) [inline apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryQueryJob.cs]: Would be good to mention that it can only be called on a completed job.

File: apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryQueryJob.cs [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+            }
+        }
+
+        // Note to reviewers: we should make it as easy as possible to do the work below, of course.
+        // Possibly BigqueryClient should have a GetResultSet(jobReference, pageSize, pageToken) call.
+
+        /// <summary>
+        /// Eagerly fetches a set of rows, up to the specified count.
+        /// </summary>
+        /// <remarks>
```
Comment (chrisdunelm) [inline apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryQueryJob.cs]: xmldoc for maxRows param?

File: apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryResultSet.cs [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        /// Schema of the results (even if there are no rows).
+        /// </summary>
+        public TableSchema Schema { get; }
+
+        /// <summary>
+        /// If non-null, indicates more results may be available. Use this page
+        /// token to fetch subsequent rows. This will never be present if fewer rows
+        /// are returned than were requested, and will never be empty.
+        /// </summary>
+        public string PageToken { get; set; }
```
Comment (chrisdunelm) [inline apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryResultSet.cs]: Does this need to be mutable?

File: apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryClient.cs (hunk 2 above)
Comment (jskeet) [inline apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryClient.cs]: Done.

File: apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryClientImpl.Queries.cs (hunk 3 above)
Comment (jskeet) [inline apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryClientImpl.Queries.cs]: Done.

File: apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryClientImpl.Queries.cs (hunk 4 above)
Comment (jskeet) [inline apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryClientImpl.Queries.cs]: Done. Oops :)

File: apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryClientImpl.Queries.cs (hunk 5 above)
Comment (jskeet) [inline apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryClientImpl.Queries.cs]: Done.

File: apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryResultSet.cs (hunk 10 above)
Comment (jskeet) [inline apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryResultSet.cs]: Nope. Fixed.

File: apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryQueryJob.cs (hunk 9 above)
Comment (jskeet) [inline apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryQueryJob.cs]: Done.

File: apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryQueryJob.cs (hunk 7 above)
Comment (jskeet) [inline apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryQueryJob.cs]: Done.

File: apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryQueryJob.cs (hunk 8 above)
Comment (jskeet) [inline apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryQueryJob.cs]: Done.

File: apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryQueryJob.cs (hunk 6 above)
Comment (jskeet) [inline apis/Google.Bigquery.V2/Google.Bigquery.V2/BigqueryQueryJob.cs]: Will do it in a separate PR.
